### ISSUE TITLE
2.0 zio-test: provide way to detect performance regressions

### ISSUE THREAD

**Issue by domdorn** [zio/zio#6684]
It would be great if we'd have a way to detect performance regressions through zio-test.
Simplest way would be to simply annotate a test like `@@ TestAspect.expectDuration(200.millis, 25.millis)` (e.g. expectedDuration, StdDeviation). 

a more advanced solution could be to tag the test with an Id, e.g. `@@ TestAspect.monitoredPerformance("JobsService-0001")`
and that would read the above values (expectedDuration, StdDeviation) from a text file checked into SCM. 

WDYT?

**Comment by adamgfraser** [comment]
Yes I would definitely like to do something like this for our own use as well as for others.

I think there are a couple of ways we could potentially approach this.

One question is whether you want to actually check times within CI and fail at that point or whether you want to track metrics of times and write them somewhere so you can monitor them over time.

Actually failing CI is obviously the strongest measure but I think it may run into issues if you have natural variations within CI so you either have to set the expectation at a very high level that is less useful or create flakiness.

The alternative of writing metrics somewhere is I think more similar to like what ZIO HTTP does with being able to request metrics for PRs or what I believe the Scala library does with being able to view metrics over time. So you could imagine ZIO Test writing metrics about all tests, including execution times and potentially other things like whether the test succeeded or failed, to a file in a structured format and then provided a tool to aggregate and visualize these outputs over time, so you could see how the duration a particular test took changed over time or from commit to commit.

**Comment by hearnadam** [comment]
I think this should be merged with #2785
